### PR TITLE
Form improvements

### PR DESCRIPTION
**Pull request by wolflu05**
This PR adds a few things which improve the form framework. This is hopefully the last PR that is needed to continue #4824. 

Specifically the following is changed:
1. A new serializer field named `DependentField` is added. To test this new serializer field, the printing plugin code at the end can be used. This is actually a sample implementation of how I imagine the machine printing to work.
2. CUI now fully supports nested object serializer fields in the form framework.
3. CUI now supports the new `DependentField`

https://github.com/inventree/InvenTree/assets/76838159/cc0eae6e-ecb8-4922-a0bc-4b51026efc11

I actually tried to implement the nested serializer field in PUI, but I sadly have not found my way around the PUI form framework code, so this would be another task.

Please let me know what you think of that implementation and if there's anything I should change.
 
---

<details><summary>Sample printing plugin code for testing</summary>

```py
from django.utils.translation import ugettext_lazy as _
from InvenTree.serializers import DependentField
from label.models import LabelTemplate
from rest_framework import serializers

# InvenTree plugin libs
from plugin import InvenTreePlugin
from plugin.mixins import LabelPrintingMixin


class SettingsTestPlugin(LabelPrintingMixin, InvenTreePlugin):
    AUTHOR = "wolflu05"
    DESCRIPTION = "Label printing test plg"
    VERSION = "0.1.1"

    NAME = "LabelPrintingTest"
    SLUG = "labelprintingtest"
    TITLE = "Label printing test PLG"

    def print_label(self, **kwargs):
        return super().print_label(**kwargs)

    def print_labels(self, label: LabelTemplate, items: list, request, **kwargs):
        return super().print_labels(label, items, request, **kwargs)

    class PrintingOptionsSerializer(serializers.Serializer):
        machine = serializers.ChoiceField(choices=[
            ("a", "Machine A"),
            ("b", "Machine B"),
            ("c", "Machine C"),
        ])

        def get_machine_options(self, data):
            if data["machine"] == "b":
                return BMachineOptionsSerializer()

        machine_options = DependentField(depends_on=["machine"], field_serializer="get_machine_options")


class BMachineOptionsSerializer(serializers.Serializer):
    select_orientation_type = serializers.CharField()
    orientation = DependentField(depends_on=["select_orientation_type"], field_serializer="get_orientation_field")

    def get_orientation_field(self, fields):
        if fields["select_orientation_type"] == "1":
            return serializers.IntegerField()
        if fields["select_orientation_type"] == "2":
            return serializers.CharField()
        if fields["select_orientation_type"] == "3":
            return serializers.ChoiceField(choices=[
                ("landscape", "Landscape"),
                ("portrait", "Portrait"),
            ])
```

</details>

### CODE REVIEW

**Comment by wolflu05** [comment]
@SchrodingersGat , @matmair this would be ready to review. I know you have a lot of other things todo with PUI, but this would be the last improvement to CUI to raise the printing options to a next level.

**Comment by SchrodingersGat** [comment]
@wolflu05 an interesting approach! Does this mean that you have to query the server on each field change, to request the new sub-field information?

**Comment by wolflu05** [comment]
Exactly @SchrodingersGat , an OPTIONS request is sent on any field change of a field that is included in the depends_on list. For text fields, a change is you edit and then click outside of the field (I used the onEdit handler for that). That way we can easily create different forms based on other values.

**Comment by SchrodingersGat** [comment]
Ok, that seems reasonable. Will you integrate support for this into PUI too?

**Comment by wolflu05** [comment]
I actually tried, but didn't got my way around. Now after I refactored the forms in #5897 I may understand the logic. If you really wish this should be done now I can see what I can do, otherwise the plan was to finish the machine integration first for what I would need this PR. But that's up to you.

**Comment by SchrodingersGat** [comment]
I'm not expecting the PUI implement straight away - but it should be kept in mind...

**Comment by SchrodingersGat** [comment]
I'm happy to merge this in for now though, thanks!

File: InvenTree/templates/js/translated/forms.js [hunk 1]
```diff
@@ -493,6 +493,30 @@ function constructFormBody(fields, options) {
         html += options.header_html;
     }
 
+    // process every field by recursively walking down nested fields
+    const processField = (name, field, optionsField) => {
+        if (field.type === "nested object") {
+            for (const [k, v] of Object.entries(field.children)) {
+                processField(`${name}__${k}`, v, optionsField.children[k]);
```
Comment (SchrodingersGat) [inline InvenTree/templates/js/translated/forms.js:500]: @wolflu05 this has broken existing form handling. e.g. try to edit an existing purchase order:

image

Please look into this one ASAP

**Comment by SchrodingersGat** [comment]
@wolflu05 please see comment above - this PR has broken some forms

**Comment by SchrodingersGat** [comment]
Looks like there are some other on-flow issues too. For now I will revert - https://github.com/inventree/InvenTree/pull/5913 - and we will need to re-asssess

**Comment by wolflu05** [comment]
Sorry that this introduced some errors, I really thought I tested with the part form and it worked. I'll definitely take a look and fix that. Good that you reverted the changes.

**Pull request merged**